### PR TITLE
Make the "Hello World" demo smaller

### PR DESCRIPTION
**Pull request by npitre**
Zephyr may look somewhat bloated by default for small apps.
Let's take the "Hello World" sample as an example for how to
reduce its size.

### CODE REVIEW

**Review by carlescufi — COMMENTED** [review]
We have `samples/basic/minimal` for minimal footprint samples. is `hello_world` a good sample to optimize on?

File: samples/hello_world/boards/qemu_cortex_m0.conf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Make this as small an example as possible
+CONFIG_MULTITHREADING=n
```
Comment (thedjnK) [inline samples/hello_world/boards/qemu_cortex_m0.conf:2]: Do not want, drivers are mostly useless without multithreading, what use is this?

**Review by thedjnK — CHANGES_REQUESTED** [review]

**Comment by npitre** [comment]
> is `hello_world` a good sample to optimize on?

The `hello_world` example is most likely the first thing anyone new to
any project will look. And in our case it is using shy of 20K of memory
by default on a Cortex-M0. That is a good way for Zephyr's first
impression to rhyme with bloatware.

File: samples/hello_world/boards/qemu_cortex_m0.conf (hunk 1 above)
Comment (npitre) [reply samples/hello_world/boards/qemu_cortex_m0.conf:2]: MCUboot is one example where thread support may not be needed and configuring it out saves a lot relatively speaking.

**Comment by kartben** [comment]
> We have `samples/basic/minimal` for minimal footprint samples. is `hello_world` a good sample to optimize on? 

Ya but `hello_world` should be pretty close to minimal IMO.
Just a wild thought but could the "minimal" trait be perhaps defined via a snippet, that both `basic/minimal` and `hello_world` would use, with `hello_world` then (re-)enabling features it really needs to run properly? But to some extent it feels that maybe this is a workaround for the fact that maybe there are some default Kconfig that we may want to review/update if they mean out-of-the-box footprint is unnecessarily bloated. Ex. I could imagine MPU becoming disabled by default, if we make the hardening tool more of a first class citizen that warns people that this might not be something they want to leave that way when rolling out in production.

File: samples/hello_world/boards/qemu_cortex_m0.conf (hunk 1 above)
Comment (nordicjm) [reply samples/hello_world/boards/qemu_cortex_m0.conf:2]: Yes, and it's disabled by default for mcuboot (which is a special case), so what use is that to an application that is going to be doing something, most likely with external hardware?

**Comment by nordicjm** [comment]
> > We have `samples/basic/minimal` for minimal footprint samples. is `hello_world` a good sample to optimize on?
> 
> Ya but `hello_world` should be pretty close to minimal IMO. Just a wild thought but could the "minimal" trait be perhaps defined via a snippet, that both `basic/minimal` and `hello_world` would use, with `hello_world` then (re-)enabling features it really needs to run properly? But to some extent it feels that maybe this is a workaround for the fact that maybe there are some default Kconfig that we may want to review/update if they mean out-of-the-box footprint is unnecessarily bloated. Ex. I could imagine MPU becoming disabled by default, if we make the hardening tool more of a first class citizen that warns people that this might not be something they want to leave that way when rolling out in production.

Disagree, `hello_world` is probably what most vendors will run right away when they add a new board or SoC and it should be instantly visible from that if something is wrong with their configuration. Hello world is not meant to be a state of the art world's smallest application, it's meant as a simple example of "does this work?"

File: samples/hello_world/boards/qemu_cortex_m0.conf (hunk 1 above)
Comment (npitre) [reply samples/hello_world/boards/qemu_cortex_m0.conf:2]: Everyone's app is its own special case. If there is justification for disablind thread support so Zephyr could be used in an extremely constrained case then why would you be against that? You are not forced to disable threads in your own project if you don't want to.

**Comment by npitre** [comment]
This PR is getting too controversial for its worth. I'm therefore closing it.
I might move its purpose to `samples/basic/minimal/` where it is arguably 
the best place for this exercise.

File: samples/hello_world/boards/qemu_cortex_m0.conf (hunk 1 above)
Comment (thedjnK) [reply samples/hello_world/boards/qemu_cortex_m0.conf:2]: > If there is justification for disablind thread support so Zephyr could be used in an extremely constrained case then why would you be against that

No, and as Carles said, that's why there is a "minimal" sample to demonstrate that

**Comment by andyross** [comment]
I missed the bikeshed here.  FWIW @npitre, I think you're absolutely right that there's a need for something in samples or docs or somewhere to treat the subject of "How to Tune Zephyr for Really Small Images", and what the various default features are that can be disabled (and what the tradeoffs of doing so are), etc...

I just don't know that hello_world is the right spot for it.  This is the app we tell people to try if they want to "Run Zephyr", and so it should reflect the most obvious/basic/vanilla/default/expected settings for everything it can.  This is the first testbed people[1] run to to try out a feature, and it's a net loss if that feature is disabled and doesn't work. 

[1] Not just new users.  I do this myself.  Daily!  And I like that it has stuff configured with reasonable defaults[2].

[2] Mostly.  The fact that it doesn't have CONFIG_ASSERT=y is a routine footgun for me.

**Comment by npitre** [comment]
The sequel is in PR #61241.

**Pull request closed without merging**